pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,681 @@
+		}
+		c.ReplaceEgressGroups(policies)
+		return nil
+	}
+	c.fullSyncWaitGroup = &c.fullSyncGroup
+
+	c.fullSyncGroup.Add(1)
+	egressInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEgressPolicy,
+			UpdateFunc: c.updateEgressPolicy,
+			DeleteFunc: c.deleteEgressPolicy,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *Controller) AddEgressGroup(group *v1beta1.EgressGroup) error {
+	c.setByGroupLock.Lock()
+	defer c.setByGroupLock.Unlock()
+	klog.Infof("%#v", group)
+	return c.addEgressGroupLocked(group)
+}
+
+func (c *Controller) PatchEgressGroup(patch *v1beta1.EgressGroupPatch) error {
+	c.setByGroupLock.Lock()
+	defer c.setByGroupLock.Unlock()
+	klog.Infof("%#v", patch)
+	return c.patchEgressGroupLocked(patch)
+}
+
+func (c *Controller) patchEgressGroupLocked(patch *v1beta1.EgressGroupPatch) error {
+	addedMembers := make([]v1beta1.GroupMember, 0, len(patch.AddedGroupMembers))
+	removedMembers := make([]v1beta1.GroupMember, 0, len(patch.RemovedGroupMembers))
+	klog.Infof("%#v", c.setByGroup)
+
+	var groupMemberSet v1beta1.GroupMemberSet
+	groupMemberSet, exists := c.setByGroup[patch.Name]
+	klog.Infof("%#v", groupMemberSet)
+	klog.Infof("%#v", exists)
+	if !exists {
+		groupMemberSet = v1beta1.GroupMemberSet{}
+	}
+	for i := range patch.AddedGroupMembers {
+		if !groupMemberSet.Has(&patch.AddedGroupMembers[i]) {
+			addedMembers = append(addedMembers, patch.AddedGroupMembers[i])
+		}
+		groupMemberSet.Insert(&patch.AddedGroupMembers[i])
+	}
+	for i := range patch.RemovedGroupMembers {
+		if groupMemberSet.Has(&patch.RemovedGroupMembers[i]) {
+			removedMembers = append(removedMembers, patch.RemovedGroupMembers[i])
+		}
+		groupMemberSet.Delete(&patch.RemovedGroupMembers[i])
+	}
+	klog.Infof("%#v", addedMembers)
+	klog.Infof("%#v", removedMembers)
+	c.setByGroup[patch.Name] = groupMemberSet
+	klog.Infof("%#v", c.setByGroup)
+	klog.Infof("%#v", c.policyMap)
+
+	c.onEgresssAdd(patch.Name, addedMembers)
+	c.onEgressRemove(removedMembers)
+	return nil
+}
+
+func (c *Controller) addEgressGroupLocked(group *v1beta1.EgressGroup) error {
+	klog.Infof("%#v", group)
+	groupMemberSet := v1beta1.GroupMemberSet{}
+	for i := range group.GroupMembers {
+		groupMemberSet.Insert(&group.GroupMembers[i])
+	}
+	oldGroupMemberSet, exists := c.setByGroup[group.Name]
+	if exists && oldGroupMemberSet.Equal(groupMemberSet) {
+		return nil
+	}
+	addedMembers := make([]v1beta1.GroupMember, 0, len(group.GroupMembers))
+	addedGroupMemberSet := groupMemberSet.Difference(oldGroupMemberSet)
+	for i := range addedGroupMemberSet {
+		addedMembers = append(addedMembers, *addedGroupMemberSet[i])
+		groupMemberSet.Insert(addedGroupMemberSet[i])
+	}
+	c.setByGroup[group.Name] = groupMemberSet
+	klog.Infof("%#v", c.setByGroup)
+
+	c.onEgresssAdd(group.Name, addedMembers)
+	return nil
+}
+
+func (c *Controller) updateEgressPolicyLocked(policy *v1alpha1.Egress) error {
+	c.policyMap[policy.Name] = policy.Spec.EgressIP
```

Comment:
There is no need to have another policyMap for egress to egress IP mapping. There is egressLister having it. This was what I meant when I suggested "just use egressLister to get egress when needed".